ols/user_data_collection/data_collector.py

```diff
@@ -0,0 +1,245 @@
+"""Collect insights and upload it to the Ingress service.
+
+It collects the data on startup and then it waits for the specified interval.
+"""
+
+import base64
+import io
+import json
+import logging
+import os
+import pathlib
+import subprocess
+import sys
+import tarfile
+import time
+
+import requests
+
+OLS_USER_DATA_PATH = os.environ.get("OLS_USER_DATA_PATH", "./user-data/")
+
+OLS_USER_DATA_COLLECTION_INTERVAL = int(
+    os.environ.get("UDC_COLLECTION_INTERVAL", 2 * 60 * 60)
+)  # 2 hours in seconds
+
+# available ingress urls are:
+# - https://console.redhat.com/ - production - works with refresh token
+# - https://console.stage.redhat.com/ - staging/dev - doesn't work with refresh token
+INGRESS_URL = os.environ.get("INGRESS_URL", "https://console.redhat.com/")
+INGRESS_ENDPOINT = os.environ.get("INGRESS_ENDPOINT", "api/ingress/v1/upload")
+INGRESS_TIMEOUT = 10  # seconds
+
+# customer portal offline token is generated from https://access.redhat.com/management/api
+CP_OFFLINE_TOKEN = os.environ.get("CP_OFFLINE_TOKEN")
+REDHAT_SSO_TIMEOUT = 5  # seconds
+
+logging.basicConfig(level=logging.DEBUG, stream=sys.stdout)
+logger = logging.getLogger(__name__)
+
+
+def refresh_token_from_offline_token(offline_token: str) -> str:
+    """Generate "refresh token" from the "offline token".
+
+    Offline token can be generated at:
+        https://access.redhat.com/management/api
+
+    Args:
+        offline_token: Offline token from the Customer Portal.
+
+    Returns:
+        Refresh token.
+    """
+    url = "https://sso.redhat.com/auth/realms/redhat-external/protocol/openid-connect/token"
+    data = {
+        "grant_type": "refresh_token",
+        "client_id": "rhsm-api",
+        "refresh_token": offline_token,
+    }
+
+    response = requests.post(url, data=data, timeout=REDHAT_SSO_TIMEOUT)
+    if response.status_code == requests.codes.ok:
+        return response.json()["access_token"]
+    else:
+        # TODO: better error handling
+        raise Exception("Failed to generate refresh token.")
+
+
+def run_oc(args: list[str]) -> subprocess.CompletedProcess:
```

Comment:
running oc commands is ok for tests.  for OLS code we need to be using a proper k8s python client lib.  we don't want to be shipping the oc binary.

Reply:
Oh man, I was happy that I finally figured out how to get the cluster id and pull secret via that (yes, had to install oc to the image).
Ok, so back to the beginning...